api/caseshare/real-api.ts

```diff
@@ -0,0 +1,160 @@
+    const {status, data}: {status: number, data: any} = await handleGet(path, req)
+    const caseUsers: CCDRawCaseUserModel[] = [...data.case_assignments]
+    const sharedCases: SharedCase[] = []
+    for (const caseUser of caseUsers) {
+      const sharedCase: SharedCase = {
+        caseId: caseUser.case_id,
+        caseTitle: caseUser.case_title,
+        sharedWith: caseUser.shared_with.map(rawSharedWith => ccdToUserDetails(rawSharedWith)),
+      }
+      sharedCases.push(sharedCase)
+    }
+    return res.status(status).send(sharedCases)
+  } catch (error) {
+    return res.status(error.status).send({
+      errorMessage: error.data,
+      errorStatusText: error.statusText,
+    })
+  }
+}
+
+export async function assignCases(req: Request, res: Response) {
+  const shareCases: SharedCase[] = req.body.sharedCases.slice()
+  const updatedSharedCases: SharedCase[] = []
+  const errorMessages: string[] = []
+  // call share case api
+  await doShareCase(req, shareCases, updatedSharedCases, errorMessages)
+  // TODO: call unshare case api
+  // await doUnshareCase(req, shareCases, updatedSharedCases)
+  const originalSharedNumber = shareCases.reduce((acc, aCase) => acc
+    + (aCase.pendingShares ? aCase.pendingShares.length : 0), 0)
+  const afterSharedNumber = updatedSharedCases.reduce((acc, aCase) => acc
+    + (aCase.pendingShares ? aCase.pendingShares.length : 0), 0)
+  // when none of the users are assigned successfully
+  if (originalSharedNumber > 0 && originalSharedNumber === afterSharedNumber) {
+    return res.status(500).send(errorMessages)
+  }
+  return res.status(201).send(updatedSharedCases)
+}
+
+async function doShareCase(req: Request, shareCases: SharedCase[],
+                           updatedSharedCases: SharedCase[],
+                           errorMessage: string[]) {
+  const path = `${ccdUrl}/case-assignments`
+  for (const aCase of shareCases) {
+    const newPendingShares = aCase.pendingShares ? aCase.pendingShares.slice() : []
+    const newSharedWith = aCase.sharedWith ? aCase.sharedWith.slice() : []
+    if (aCase.pendingShares) {
+      for (const pendingShare of aCase.pendingShares) {
+        const payload = {
+          'assignee_id': pendingShare.idamId,
+          'case_id': aCase.caseId,
+          'case_type_id': aCase.caseTypeId,
+        }
+        logger.info('Request payload:', JSON.stringify(payload))
+        try {
+          const {status}: { status: number } = await handlePost(path, payload, req)
+          if (status === 201) {
+            newSharedWith.push(pendingShare)
+            newPendingShares.splice(newPendingShares.findIndex(iShare => iShare.idamId === pendingShare.idamId), 1)
+          }
+        } catch (error) {
+          logger.error('Error message:', JSON.stringify(error.data))
+          errorMessage.push(`${error.status} ${error.statusText} ${JSON.stringify(error.data)}`)
+        }
+      }
+    }
+    const newSharedCase = {
```

Comment:
This is not good, we're side-effecting out-side of the node function, how do we know when this function has completed it's work? This function should be 'returning'.